### PR TITLE
#21 UC Secure Fischlin Proofs

### LINKED ISSUE #21: Implement Fischlin's Transform for UC-Secure ZK Proofs
There are cases where the security proof requires a UC-secure ZK proof, for which the Fischlin transformation is typically used. Sections 2.2 and 2.3 in [Chen and Lindell (2024)](https://eprint.iacr.org/2024/526.pdf) describe protocols for an optimized implementation. Since we already implemented support for parallel repetitions, most of the heavy lifting has already been done. We are left with:
1. calculating `common-h` in 2.2.3 (actually we can re-use the existing calculate challenges, just take a single challenge; note that we need `2k` i.e. 256-bit challenges, not 128-bit like existing code).
2. in 2.2.4. can simply do an infinite loop. Instead of the existing challenges, we use an incrementing challenge. We don't multiply by the challenge, since we are using incrementing ones we can just add `w` at each iteration. We break if the hash is 0 in the first bits. This is the main change to the prover.
3. We need to also send the challenges vector `e` in the proof. Perhaps it could be an empty vector if not Fischlin.

In the verifier, the only change we need to add is that the challenges are not computed, instead we check that each challenge yields starting with 0 bits.

Optionally, batch verification from section 2.3 could be implemented in the future to save `4x` in time. Currently it is not planned as for the language we need (knowledge of decommittment) this should be optimal enough even without this optimization and would not be the bottleneck.

### PR DESCRIPTION
**Pull request by ycscaly**
Resolves #21

### CODE REVIEW

File: src/proof/fischlin.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    // Number of parallel repetitions required to get a negligible soundness error.
+    const REPETITIONS: usize,
+    // The language we are proving
+    Language: language::Language<REPETITIONS>,
+    // A struct used by the protocol using this proof,
+    // used to provide extra necessary context that will parameterize the proof (and thus verifier
+    // code) and be inserted to the Fiat-Shamir transcript
+    ProtocolContext: Clone,
+> {
+    pub maurer_proof: super::Proof<REPETITIONS, Language, ProtocolContext>,
```
Comment (dolevmu) [inline src/proof/fischlin.rs:27]: Yes proof should have REPETITIONS, Language I think shouldn't.

File: src/proof/fischlin.rs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    pub challenges: [ComputationalSecuritySizedNumber; REPETITIONS],
+}
+
+impl<
+        const REPETITIONS: usize,
+        Language: language::Language<REPETITIONS>,
+        ProtocolContext: Clone + Serialize,
+    > Proof<REPETITIONS, Language, ProtocolContext>
+{
+    /// Prove a Universally Composable (UC) Maurer zero-knowledge claim via Fischlin's transform.
```
Comment (dolevmu) [inline src/proof/fischlin.rs:39]:
```suggestion
    /// Prove a Universally Composable (UC) Extractable zero-knowledge (ZK) Maurer statement via Fischlin's transform.
```
Just emphasizing that we have a UC extractor, that's the main issue.

File: src/proof/fischlin.rs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    /// Returns the zero-knowledge proof.
+    pub fn prove(
+        protocol_context: &ProtocolContext,
+        language_public_parameters: &Language::PublicParameters,
+        witness: Language::WitnessSpaceGroupElement,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<(Self, Language::StatementSpaceGroupElement)> {
+        if REPETITIONS == 0 || REPETITIONS > ComputationalSecuritySizedNumber::BITS {
+            return Err(Error::UnsupportedRepetitions);
+        }
```
Comment (dolevmu) [inline src/proof/fischlin.rs:50]: I think `challenge_bits` does this for you already right?

File: src/proof/fischlin.rs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        ProtocolContext: Clone + Serialize,
+    > Proof<REPETITIONS, Language, ProtocolContext>
+{
+    /// Prove a Universally Composable (UC) Maurer zero-knowledge claim via Fischlin's transform.
+    /// Implements [Chen and Lindell (2024)](https://eprint.iacr.org/2024/526.pdf), sections 2.2, 2.3.
+    /// Returns the zero-knowledge proof.
+    pub fn prove(
+        protocol_context: &ProtocolContext,
+        language_public_parameters: &Language::PublicParameters,
+        witness: Language::WitnessSpaceGroupElement,
```
Comment (dolevmu) [inline src/proof/fischlin.rs:45]:
```suggestion
        witness: Language::WitnessSpaceGroupElement, // $w$
```

File: src/proof/fischlin.rs [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<(Self, Language::StatementSpaceGroupElement)> {
+        if REPETITIONS == 0 || REPETITIONS > ComputationalSecuritySizedNumber::BITS {
+            return Err(Error::UnsupportedRepetitions);
+        }
+
+        let statement = Language::homomorphose(&witness, language_public_parameters)?;
```
Comment (dolevmu) [inline src/proof/fischlin.rs:52]:
```suggestion
        let statement = Language::homomorphose(&witness, language_public_parameters)?; // $x$
```

File: src/proof/fischlin.rs [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+        let (randomizers, statement_masks) =
```
Comment (dolevmu) [inline src/proof/fischlin.rs:54]:
```suggestion
        let (randomizers, statement_masks) = // $(\vec{\sigma}, \vec{m_i})
```

File: src/proof/fischlin.rs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+            super::Proof::<REPETITIONS, Language, ProtocolContext>::sample_randomizers_and_statement_masks(language_public_parameters, rng)?;
+
+        let statement_masks_values =
+            Language::StatementSpaceGroupElement::batch_normalize_const_generic(statement_masks);
+
+        // $common-h$
```
Comment (dolevmu) [inline src/proof/fischlin.rs:60]: 👍

File: src/proof/fischlin.rs [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+            &statement_masks_values,
+        )?;
+
+        let challenges_and_responses: [Result<(
+            ComputationalSecuritySizedNumber,
+            WitnessSpaceValue<REPETITIONS, Language>,
+        )>; REPETITIONS] = array::from_fn(|i| {
+            let randomizer = randomizers[i].clone();
+
+            let mut challenge = ComputationalSecuritySizedNumber::ZERO;
```
Comment (dolevmu) [inline src/proof/fischlin.rs:74]:
```suggestion
            let mut challenge = ComputationalSecuritySizedNumber::ZERO; // $e_i$
```

File: src/proof/fischlin.rs [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+            let mut response = randomizer;
```
Comment (dolevmu) [inline src/proof/fischlin.rs:75]:
```suggestion
            let mut response = randomizer; // $z_i$
```

File: src/proof/fischlin.rs [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+
+        let uc_proof = Self {
+            maurer_proof,
+            challenges,
+        };
+
+        Ok((uc_proof, statement))
+    }
+
+    /// Verify a Universally Composable (UC) Maurer zero-knowledge claim via Fischlin's transform.
```
Comment (dolevmu) [inline src/proof/fischlin.rs:114]:
```suggestion
    /// Verify a Universally Composable (UC) Extractable Maurer zero-knowledge claim via Fischlin's transform.
```

File: src/proof/fischlin.rs [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+        transcript.append_uint(b"challenge", &challenge);
+        // Add the response $z_i$
+        let res = transcript.serialize_to_transcript_as_json(b"response", &response);
+        if res.is_err() {
+            return Err(res.err().unwrap())?;
+        }
+        // Finalize the hash computation to get $h_i = H(common-h, i, e_i, z_i)$.
+        let hash: ComputationalSecuritySizedNumber = transcript.challenge(b"hash");
+
+        // TODO: better name?
```
Comment (dolevmu) [inline src/proof/fischlin.rs:198]: In proof-of-work blockchains, the number of leading zeros you want at the output of the hash of your block is called the `difficulty target`. I think it's a good fit.

File: src/proof/fischlin.rs [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+        // Compute $b$ such that $b \cdot \rho >= \kappa$
```
Comment (dolevmu) [inline src/proof/fischlin.rs:199]:
```suggestion
        // Compute the difficulty target $b$ such that $b \cdot \rho >= \kappa$.
        // This determines the number of tralining zeros in H_b(common_h, i, e_i, z_i).
```

File: src/proof/fischlin.rs [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+        let hash_zero_bits =
+            (ComputationalSecuritySizedNumber::BITS + (REPETITIONS - 1)) / REPETITIONS;
+
+        // TODO: verify correctness
+        Ok(
+            (hash & ComputationalSecuritySizedNumber::from((1u64 << hash_zero_bits) - 1))
```
Comment (dolevmu) [inline src/proof/fischlin.rs:205]: It's clearly true unless your prover does O(2^64) work, but you may add a check that `2^64 > b`.
Given that, yes this works.

Alternatively, doesn't this work?
```suggestion
            (hash & ((ComputationalSecuritySizedNumber::from(1u64) << hash_zero_bits) - 1))
```

**Review by dolevmu — COMMENTED** [review]
Great! Consider my comments and let ping for final approval.

File: src/proof/fischlin.rs [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+use crypto_bigint::rand_core::CryptoRngCore;
+use crypto_bigint::Concat;
+use group::{ComputationalSecuritySizedNumber, GroupElement};
+use merlin::Transcript;
+use proof::TranscriptProtocol;
+use serde::{Deserialize, Serialize};
+use std::array;
+
+/// A Universally Composable (UC) Maurer Zero-Knowledge Proof via Fischlin's transform.
+/// Implements [Chen and Lindell (2024)](https://eprint.iacr.org/2024/526.pdf).
```
[the remark this answers is not in the document]
Comment (ycscaly) [reply src/proof/fischlin.rs:15]: Nah, already in the issue I explained something like that

File: src/proof/fischlin.rs (hunk 1 above)
Comment (ycscaly) [reply src/proof/fischlin.rs:27]: Maybe that's a better design, but all my code already works this way - we can create an issue for that in the future

File: src/proof/fischlin.rs (hunk 3 above)
Comment (ycscaly) [reply src/proof/fischlin.rs:50]: Yes, but we don't use it in `prove`

File: src/proof/fischlin.rs (hunk 11 above)
Comment (ycscaly) [reply src/proof/fischlin.rs:198]: "The difficulty is just a measure of how much the target has moved from its initial starting value. Or in other words, how much more difficult it is to mine a block compared to when the blockchain first started."

Comment (ycscaly) [reply src/proof/fischlin.rs:198]: "The [target](https://learnmeabitcoin.com/technical/mining/target/) is the actual limbo pole that block hashes have to get below for a new block to be added on the blockchain."

File: src/proof/fischlin.rs [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+                vec![statement.value()],
+                statement_masks_values,
+            )?;
+
+        Ok(transcript.challenge(b"common hash"))
+    }
+
+    /// Performs the Fischlin transformation check that $h_i = H(common-h, i, e_i, z_i)$ hits the target,
+    /// i.e. it starts with `target_bits` $b$ zeros.
+    fn hash_hits_target(
```
Comment (ycscaly) [inline src/proof/fischlin.rs:177]:
```suggestion
    fn is_target(
```
@dolevmu which name is more suitable?

**Review by ofirfriedman42 — APPROVED** [review]

File: src/proof/fischlin.rs (hunk 11 above)
Comment (dolevmu) [reply src/proof/fischlin.rs]: Okay cool, we have a name then 👍

File: src/proof/fischlin.rs (hunk 15 above)
Comment (dolevmu) [reply src/proof/fischlin.rs:177]: I like `hash_hits_target`. Also your input to this function is the input to the hash H_b, so you don't ask if the input **is** the target, but if the **hash of it is** the target. So at least `hash_is_target` but hits sounds better.

**Review by dolevmu — APPROVED** [review]

**Pull request merged**